tiny_dnn/core/framework/tensor.h

```diff
@@ -0,0 +1,281 @@
+    // from one device to another.
+    void toDevice(const Device& device) {
+#if defined(USE_OPENCL) || defined(USE_CUDA)
+        CLCudaAPI::Context ctx = device.context();
+        CLCudaAPI::Queue queue = device.queue();
+
+        device_data_ = buffer_ptr(new buffer(
+            ctx, queue, host_data_->begin(), host_data_->end()));
+#endif
+    }
+
+    // this is only a proof of concept to copy data
+    // from one device to another.
+    void fromDevice(const Device& device) {
+#if defined(USE_OPENCL) || defined(USE_CUDA)
+	CLCudaAPI::Queue queue = device.queue();
+
+        device_data_->Read(
+	    queue, host_data_->size(), &host_data_->at(0));
+#endif
+    }
+
+    /*template<typename T, typename U = float_t>
+    T host_data() const {
+        return static_cast<T>(*access_data(0));
+    }
+
+    template<typename T, typename U = float_t>
+    T mutable_host_data() {
+        return static_cast<T>(*access_data(0));
+    }
+
+    template<typename T, typename U = float_t>
+    T device_data() const {
+	fromDevice(device_);
+        return static_cast<T>(*access_data(0));
+    }
+
+    template<typename T, typename U = float_t>
+    T mutable_device_data() {
+	fromDevice(device_);
+        return static_cast<T>(*access_data(0));
+    }*/
+
+    size_t size() const {
+        size_t new_size = 1;
+        for (auto d : shape_) { new_size *= d; }
+        return new_size;
+    }
+
+ private:
+    // Initializes the data buffer with zeroes
+    void resize(const U value = 0) {
+        if (!host_data_) {
+            host_data_ = std::unique_ptr<std::vector<U> >(
+                new std::vector<U>(size(), value));
+        } else {
+            host_data_->resize(size(), value);
+        }
+    }
+
+    // Initializes the shape vector
+    void reshape(const cnn_size_t d0,
+                 const cnn_size_t d1,
+                 const cnn_size_t d2,
+                 const cnn_size_t d3) {
+	    shape_.resize(4);
+	    shape_[0] = d0;
+	    shape_[1] = d1;
+	    shape_[2] = d2;
+	    shape_[3] = d3;
+    }
+
+    // Method to access to the tensor data.
+    // It checks if the requested position is feasible or not.
+    template<typename T>
+    T* access_data(const cnn_size_t d0,
+                   const cnn_size_t d1,
+                   const cnn_size_t d2,
+                   const cnn_size_t d3) const {
+        if (d0 > shape_[0] || d1 > shape_[1] ||
+            d2 > shape_[2] || d3 > shape_[3]) {
+            throw nn_error("Access tensor out of range.");
+        }
+
+        U* value = &host_data_->at(shape_[1] * shape_[2] *
+            ( shape_[3] * d0 + d3 ) + d1 + d2);
+
+        // in case that requested type is not the same as
+        // the specified during the tensor initilization
+        // we cast the type.
+        if (!std::is_same<T,U>::value) {
+            return reinterpret_cast<T*>(value);
+        }
+        return value;
+    }
+
+    template<typename T>
+    T* access_data(const size_t index) const {
```

Comment:
Maybe I'm missing something but it seems to me that type of return value of this method is `T*`, I simply thought it should be `const T*`. And you also need to make non-const version of this method.
```
template<typename T>
T* access_data(const size_t index) {
    // almost same impl..
}
```